File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.py
def solution(stones, k):
    def check(n):
        jump = 0
        
        for stone in stones:
            if stone - n >= 0:
                jump = 0
            elif (jump := jump + 1) >= k:
                return False
        
        return True
    
    
    answer = 0
    left, right = 0, max(stones)
    
    while left <= right:
        mid = (left + right) // 2
        
        if check(mid):
            answer = max(answer, mid)
            left = mid + 1
        else:
            right = mid - 1
        
    return answer

File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.java
class Solution {

    private int[] stones;
    private int k;

    public int solution(int[] stones, int k) {
        this.stones = stones;
        this.k = k;

        int left = 0, right = 0;
        for (int s : stones) right = Math.max(right, s);

        int answer = 0;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (check(mid)) {
                answer = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return answer;
    }

    private boolean check(int n) {
        int jump = 0;
        for (int stone : stones) {
            if (stone - n >= 0) {
                jump = 0;
            } else if (++jump >= k) {
                return false;
            }
        }
        return true;
    }
}